pkg/queue/queue.go

```diff
@@ -0,0 +1,119 @@
+package queue
+
+import (
+	"container/list"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/stackrox/rox/pkg/concurrency"
+	"github.com/stackrox/rox/pkg/logging"
+	"github.com/stackrox/rox/pkg/metrics"
+	"github.com/stackrox/rox/pkg/sync"
+)
+
+var (
+	log = logging.LoggerForModule()
+)
+
+// Queue provides a thread-safe queue for type T.
+// The queue allows to push, pull, and blocking pull.
+// Additionally, it exposes safety guards such as a max size as well as metrics to track the queue growth and size.
+type Queue[T comparable] struct {
+	maxSize        int
+	counterMetric  *prometheus.CounterVec
+	queue          *list.List
+	notEmptySignal concurrency.Signal
+	mutex          sync.Mutex
+}
+
+// OptionFunc provides options for the queue.
+// Note that the comparable type is currently required, once we upgrade to go1.20 we can switch this to
+// any and creation will be much easier.
+type OptionFunc[T comparable] func(queue *Queue[T])
+
+// WithCounterVec provides a counter vec which tracks added and removed items from the queue.
+func WithCounterVec[T comparable](vec *prometheus.CounterVec) OptionFunc[T] {
+	return func(queue *Queue[T]) {
+		queue.counterMetric = vec
```

Comment:
Have you considered using a callback function a la `increaseQueueSizeCounter()`, instead of passing a Prometheus object directly? Seems a bit cleaner to me because then we don't have any `prometheus` in the queue module.

Reply:
Hmm, I did initially, but the `label` would require importing `metrics` anyways for `Remove/Add`, so given that it felt cleaner to me to just include the complete counter in the package, providing a more clear and cleaner API to clients of the package.